adapters/bidscube/bidscube.go

```diff
@@ -0,0 +1,124 @@
+package bidscube
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strconv"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, _ *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	impressions := request.Imp
+	result := make([]*adapters.RequestData, 0, len(impressions))
+	var errs []error
+
+	for _, impression := range impressions {
+		var impExt map[string]json.RawMessage
+		if err := json.Unmarshal(impression.Ext, &impExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		bidderExt, bidderExtExists := impExt["bidder"]
+		if !bidderExtExists || len(bidderExt) == 0 {
+			errs = append(errs, errors.New("bidder parameters required"))
+			continue
+		}
+
+		impression.Ext = bidderExt
+		request.Imp = []openrtb2.Imp{impression}
+		body, err := json.Marshal(request)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		result = append(result, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     a.endpoint,
+			Body:    body,
+			Headers: headers,
+		})
+	}
+
+	request.Imp = impressions
+	return result, errs
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		return nil, []error{&errortypes.BadInput{
+			Message: "unexpected status code: " + strconv.Itoa(responseData.StatusCode),
+		}}
+	case http.StatusOK:
+		break
+	default:
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "unexpected status code: " + strconv.Itoa(responseData.StatusCode),
+		}}
+	}
+
+	var bidResponse openrtb2.BidResponse
+	err := json.Unmarshal(responseData.Body, &bidResponse)
+	if err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	response := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+
+	for _, seatBid := range bidResponse.SeatBid {
+		for i := range seatBid.Bid {
+			var bidExt map[string]interface{}
+			if err := json.Unmarshal(seatBid.Bid[i].Ext, &bidExt); err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			bidType := bidExt["prebid"].(map[string]interface{})["type"].(string)
+			response.Bids = append(response.Bids, &adapters.TypedBid{
```

Comment:
That was quick. Thank you.

I believe this code introduces a chance of a panic if the `map[string]interface{}` or `string` type conversions fail. I think it would be better to use the jsonparser package here. We often prefer it for pulling out a single value from a json blob.

```
bidType, err = jsonparser.GetString(seatBid.Bid[i].Ext, "prebid", "type")
if err == nil {
	errs = append(errs, err)
	continue
}
```

Reply:
Thanks, done